### ISSUE TITLE
OutOfMemoryError for layer calculation when max index=22

### ISSUE THREAD

**Issue by vikulin** [iotaledger/compass#121]
### Bug description
OutOfMemoryException for layer calculation when max index=22 in config.json

### Hardware Spec
6GB RAM, Ubuntu 18.0.1 x86_64

### Steps To Reproduce
1. Set max milestone index=22 in config.json
2. Run ./01_calculate_layers.sh

### Expected behaviour
Layer is calculated

### Actual behaviour
OOM exception

### Errors

<img width="398" alt="OOM" src="https://user-images.githubusercontent.com/743622/58688590-92eace80-838d-11e9-875a-48c60568bd9c.png">

**Comment by karimodm** [comment]
@vikulin I am unable to reproduce what you are experiencing. Can you please post the `config.json` that you are using? Also: there is no such `max milestone index` relevant option for layers generation.

**Comment by vikulin** [comment]
posted

пн, 3 июн. 2019 г. в 13:20, Andrea V <notifications@github.com>:

**Comment by vikulin** [comment]
My config.json:

{
  "seed": "HTCRDHFEYGCZEKMVZPPZJFOPSCMYMHRITIMAIBYMHXESUZWTMNGMRLZNGAWILELGGUDJWVNNKOPMBAAHKD",
  "powMode": "CURLP81",
  "sigMode": "CURLP27",
  "security": 1,
  "depth": 24,
  "milestoneStart": 0,
  "mwm": 9,
  "tick": 60000,
  "host": "http://localhost:14265"
}

**Comment by karimodm** [comment]
TL;DR you should not generate such a big tree (31 years at your tick rate) on a such small machine.

Incrementing the Merkle tree's depth increases exponentially the amount of RAM required to hold the Merkle tree leaves, subsequently it will also need that information in order to build the tree up to the root. 6GB of RAM are not enough to compute a tree with almost 20 million leaves. Unfortunately I am not able to tell how much RAM you would need exactly to calculate it. 
In any case here are some guidelines: we benchmarked KERL `sigMode` with depth 23 and security 2. The generation took around 48 hours on a 64GB / 32 cores machine. Please note that security level 2 almost doubles the complexity of the subseed leaves derivation of the tree. If you issue a milestone every 60 seconds, the tree you are trying to generate above is gonna last you a bit more than 31 years. So either reduce the tree size or compute on a bigger machine.

**Comment by vikulin** [comment]
Yes, I tryed 22 number and your script need to be fixed as follow adding
-Xmx4G option:
vadym@vm:~/compass/docs/private_tangle$ cat 01_calculate_layers.sh
#!/bin/bash

scriptdir=$(dirname "$(readlink -f "$0")")
. $scriptdir/lib.sh

load_config

docker run -t --rm -v $scriptdir/data:/data
iota/compass/docker:layers_calculator "-Xmx4G" layers_calculator_deploy.jar
....


вт, 4 июн. 2019 г. в 13:38, Andrea V <notifications@github.com>:

**Comment by karimodm** [comment]
The change you are suggesting won't prevent the JVM from OOMing: it will prevent the JVM from OOMing the OS. In addition, since we cannot reliably calculate how much memory the JVM needs for each config.json tree depth / algorithm / security we cannot set an upper memory bound for the JVM: if we are too convervative the tree generation will fail mid-flight even if the OS had enough memory, if we are too generous it is like not setting it at all and leave the job to the OS.

**Comment by vikulin** [comment]
You can add such parameter to be configurable: BTW now everyone can fix it
playing with the JVM option.

вт, 4 июн. 2019 г. в 13:59, Andrea V <notifications@github.com>:
